File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        nodes = []
        
        def inOrderTraversal(root):
            if not root:
                return
            
            if not root.left and not root.right:
                nodes.append(root.val)
                return

            inOrderTraversal(root.left)
            nodes.append(root.val)
            inOrderTraversal(root.right)
        
        inOrderTraversal(root)

        for i in range(1, len(nodes)):
            if nodes[i - 1] >= nodes[i]:
                return False
        
        return True

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


/**
Solution Approach: Lets do this in a BFS style. 

Start off by adding the root in its own vector. OH NO NEVER MIND!
Create a queue. Add the root to the queue. Then, check size of queue. 
This is how many elements need to be added. Add the first X elements 
to a vector. Push this vector onto the solution.
**/

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        std::vector<vector<int>> sol;

        if (root == nullptr) {
            return sol;
        }

        std::queue<TreeNode*> toVisit;
        toVisit.push(root);
        
        while (!toVisit.empty()) {
            std::vector<int> nodes;
            auto n = toVisit.size();
            for (auto i = 0; i < n; i++) {
                auto node = toVisit.front();
                toVisit.pop();

                if (node->left != nullptr) {
                    toVisit.push(node->left);
                }

                if (node->right != nullptr) {
                    toVisit.push(node->right);
                }

                nodes.push_back(node->val);
            }
            sol.push_back(nodes);
        }

        return sol;
    }
};
